include/dlaf/util_math.h

```diff
@@ -10,17 +10,29 @@
 
 #pragma once
 
+#include "util_macros.h"
+
 /// @file
 
 namespace dlaf {
 namespace util {
 
-/// @brief Returns @c ceiling(@p num / @p den) for integer types.
+/// @brief Returns ceiling(num/den) for integer types.
+///
+/// It accepts both signed and unsigned integer types, but as precondition states both values must
+/// not be negative (additionally, @p den must not be zero either)
+///
+/// @tparam IntNumType and IntDenType have to be integer types.
+/// @param num
+/// @param den
+/// @return ceiled division of type @a IntNumType
 ///
-/// @tparam IntType has to be an integer type
-/// @pre @a num >= 0 and @a den >= 0
-template <class IntType>
-constexpr IntType ceilDiv(IntType num, IntType den) {
+/// @pre @a num >= 0 and @a den > 0.
+/// @pre @a num + @a den - 1 must not overflow @a IntNumType max value
+template <typename IntNumType, typename IntDenType>
+constexpr
+RETURN_TYPE_IF(IntNumType, std::is_integral<IntNumType>::value && std::is_integral<IntDenType>::value)
```

Comment:
Why return type is `IntNumType`? What about `auto`?

Reply:
Here SFINAE is used with `std::enable_if` to constraint parameters to be integers.

We cannot use `auto` because it cannot be a template parameter, in fact this is not allowed by the language
```cpp
std::enable_if_t<std::is_integral<IntNumType>::value && std::is_integral<IntDenType>::value, auto>
```